### PR TITLE
Add MPR121 support for esphome-core

### PR DESCRIPTION
**Pull request by mvturnho**
## Description:

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome](https://github.com/esphome/esphome) with python changes (if applicable):** esphome/esphome#<esphome PR number goes here>
**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Checklist:

  - [x] The code change is tested and works locally.
  - [x] The code change follows the [standards](https://esphome.io/guides/contributing.html#contributing-to-esphome-core)

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: src/esphome/application.h [hunk 1]
```diff
@@ -841,6 +842,9 @@ class Application {
   sensor::APDS9960 *make_apds9960(uint32_t update_interval = 60000);
 #endif
 
+#ifdef USE_MPR121
+  sensor::MPR121_Sensor *make_mpr121( uint8_t address = 0x5A);
```
Comment (OttoWinter) [inline src/esphome/application.h]:
```suggestion
  sensor::MPR121_Sensor *make_mpr121(uint8_t address = 0x5A);
```

File: src/esphome/sensor/mpr121_sensor.h [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+#include "esphome/defines.h"
+
+#ifdef USE_MPR121
+
+#include "esphome/sensor/sensor.h"
+#include "esphome/binary_sensor/binary_sensor.h"
+#include "esphome/i2c_component.h"
+#include "esphome/component.h"
+
+#define MPR121_I2CADDR_DEFAULT 0x5A
```
Comment (OttoWinter) [inline src/esphome/sensor/mpr121_sensor.h]: Do not use defines (discouraged by google C++ guide and good practice)

```suggestion
static const uint8_t MPR121_DEFAULT_I2C_ADDR = 0x5A;
```

Or... I would just remove it entirely (in application.cpp no const is used, so having it here doesn't do anything)

File: src/esphome/sensor/mpr121_sensor.h [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        MPR121_GPIOSET      = 0x78,
+        MPR121_GPIOCLR      = 0x79,
+        MPR121_GPIOTOGGLE   = 0x7A,
+
+        MPR121_SOFTRESET    = 0x80,
+    };
+
+    class MPR121_Channel : public binary_sensor::BinarySensor {
+        public:
+            int channel = 0;
```
Comment (OttoWinter) [inline src/esphome/sensor/mpr121_sensor.h]: Instance attributes should be protected and end with `_` (see contributing).
```cpp
  void process_(...);

protected:
  int channel_{0};
```

File: src/esphome/sensor/mpr121_sensor.h [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+            MPR121_Channel(const std::string &name,int channel_num = 0);
+            void process_(uint16_t *data, uint16_t *last_data);
+    };
+
+    class MPR121_Sensor : public Component , public I2CDevice {
+        public:
+            MPR121_Sensor(I2CComponent *parent, uint8_t address = MPR121_I2CADDR_DEFAULT);
+            sensor::MPR121_Channel *add_channel(sensor::MPR121_Channel *channel);
+            void process_(uint16_t *data,uint16_t *last_data);
```
Comment (OttoWinter) [inline src/esphome/sensor/mpr121_sensor.h]:
```suggestion
            void process_(uint16_t *data, uint16_t *last_data);
```

File: src/esphome/sensor/mpr121_sensor.h [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            void setup() override;
+            void dump_config() override;
+            //float get_setup_priority() const override;
```
Comment (OttoWinter) [inline src/esphome/sensor/mpr121_sensor.h]: Please do override `get_setup_priority` explicitly.

File: src/esphome/sensor/mpr121_sensor.h [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+            void loop() override;
+        protected:
+            std::vector<MPR121_Channel *> channels{};
```
Comment (OttoWinter) [inline src/esphome/sensor/mpr121_sensor.h]: Class variables should end in `_`

```suggestion
            std::vector<MPR121_Channel *> channels_{};
```

File: src/esphome/sensor/mpr121_sensor.cpp [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    uint16_t MPR121_Sensor::read_mpr121_channels() {
+        uint16_t val = 0;
+        this->read_byte_16(MPR121_TOUCHSTATUS_L, &val);
+        uint8_t lsb = val >> 8;
+        uint8_t msb = val ;
+        val = ((uint16_t) msb) << 8;
+        val |= lsb;
+
+        //ESP_LOGD(TAG,"msb:%d lsb:%d value: %d",msb,lsb,val);
+        return val ;
```
Comment (OttoWinter) [inline src/esphome/sensor/mpr121_sensor.cpp]:
```suggestion
        return val;
```

File: src/esphome/sensor/mpr121_sensor.cpp [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    MPR121_Channel *MPR121_Sensor::add_channel(sensor::MPR121_Channel* channel) {
+      this->channels.push_back(channel);
+      return channel;
+    }
+
+    void MPR121_Sensor::process_( uint16_t *data, uint16_t *last_data) {
+      for (auto *channel : this->channels) {
+        channel->process_(data,last_data);
```
Comment (OttoWinter) [inline src/esphome/sensor/mpr121_sensor.cpp]:
```suggestion
        channel->process_(data, last_data);
```

File: src/esphome/sensor/mpr121_sensor.cpp [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+                ESP_LOGE(TAG, "Communication with MPR121 failed!");
+                break;
+            case WRONG_CHIP_STATE:
+                ESP_LOGE(TAG, "MPR121 has wrong default value for CONFIG2?");
+                break;
+            case NONE:
+            default:
+                break;
+        }
+        ESP_LOGCONFIG(TAG, "  currvalue: %d", this->read_mpr121_channels());
```
Comment (OttoWinter) [inline src/esphome/sensor/mpr121_sensor.cpp]: `dump_config` should not do any I/O. Please remove this.

File: src/esphome/sensor/mpr121_sensor.cpp [hunk 10]
```diff
@@ -0,0 +1,125 @@
+#include "esphome/defines.h"
+#define USE_MPR121 1
```
Comment (OttoWinter) [inline src/esphome/sensor/mpr121_sensor.cpp]: Remove this - it is already set in defines.h (and ESPHome will set it with BUILD_FLAGS)

**Review by OttoWinter — COMMENTED** [review]
- Please do not indent everything in the .cpp files, see other files for proper indentation (or run clang-format on the file)
- This is only a binary sensor, right? Then this should go in the binary_sensor folder and namespace.
- No commented out code please - If you think some lines are useful for debugging you can use the `ESP_LOGV` or `ESP_LOGVV` macros, but no commented out code please.

**Comment by mvturnho** [comment]
@OttoWinter corrected all issues and tested on esp8266 and esp32.
For esp32 you need the arduino latest dev version.

```yaml
esphome:
  name: esp32mpr121
  platform: ESP32
  board: nodemcu-32s
  arduino_version: dev
```

File: src/esphome/binary_sensor/mpr121_sensor.h [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+  MPR121_TARGETLIMIT = 0x7F,
+  MPR121_GPIODIR = 0x76,
+  MPR121_GPIOEN = 0x77,
+  MPR121_GPIOSET = 0x78,
+  MPR121_GPIOCLR = 0x79,
+  MPR121_GPIOTOGGLE = 0x7A,
+  MPR121_SOFTRESET = 0x80,
+};
+
+class MPR121_Channel : public binary_sensor::BinarySensor {
```
Comment (OttoWinter) [inline src/esphome/binary_sensor/mpr121_sensor.h]:
```suggestion
class MPR121Channel : public binary_sensor::BinarySensor {
```

class names are `UpperCamelCase`

File: src/esphome/binary_sensor/mpr121_sensor.h [hunk 12]
```diff
@@ -0,0 +1,91 @@
+#ifndef ESPHOME_SENSOR_MPR121_H
+#define ESPHOME_SENSOR_MPR121_H
```
Comment (OttoWinter) [inline src/esphome/binary_sensor/mpr121_sensor.h]:
```suggestion
#define ESPHOME_BINARY_SENSOR_MPR121_H
```

File: src/esphome/defines.h [hunk 13]
```diff
@@ -214,6 +215,15 @@
   #endif
 #endif
 
+#ifdef USE_MPR121
+  #ifndef USE_SENSOR
+    #define USE_SENSOR
+  #endif
```
Comment (OttoWinter) [inline src/esphome/defines.h]: Remove `#ifndef USE_SENSOR` block

File: src/esphome/binary_sensor/mpr121_sensor.cpp [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+namespace binary_sensor {
+
+static const char *TAG = "binary_sensor.mpr121";
+
+MPR121_Channel::MPR121_Channel(const std::string &name, int channel_num) : BinarySensor(name) {
+  channel_ = channel_num;
+}
+
+void MPR121_Channel::process(uint16_t *data, uint16_t *last_data) {
+  if ((*data & _BV(channel_)) && !(*last_data & _BV(channel_))) {
```
Comment (OttoWinter) [inline src/esphome/binary_sensor/mpr121_sensor.cpp]: What is `_BV`? If it is a bit shift, use `(1 << this->channel_)` instead.

**Review by OttoWinter — COMMENTED** [review]
Do you know why you need dev arduino version for ESP32? I don't see anything in the code that would require it. If it is because of i2c, try different i2c frequencies (I don't like telling users to use arduino dev versions in the docs, since some of them are incredibly unstable and may break at any moment in time)

**Comment by mvturnho** [comment]
@OttoWinter I think I fixed all issues. Thanks for your patience so far.

I tested some further with the esp32 and nothing but the latest dev version for arduino esp32 seems to solve the error. 
I tried:
- other GPIO pins for I2C
- other device setup sequence from datasheets
- different bus frequencies 10khz up to 400khz
- disconnect the device, reset esp32 and connect device then another reset (gues I was deperate ;-)

On esp8266 no issue whatsoever. 

I ordered some other mpr121 breakout boards (I have only one for testing) maybe my device has some hardware issue.

Error on esp32
```
[11:40:00][I][i2c:048]: Scanning i2c bus for active devices...
[11:40:00][E][esp32-hal-i2c.c:1146] i2cCheckLineState(): Bus Invalid State, TwoWire() Can't init
[11:40:00][E][esp32-hal-i2c.c:1146] i2cCheckLineState(): Bus Invalid State, TwoWire() Can't init
[11:40:00][E][esp32-hal-i2c.c:1146] i2cCheckLineState(): Bus Invalid State, TwoWire() Can't init
[11:40:00][E][esp32-hal-i2c.c:1146] i2cCheckLineState(): Bus Invalid State, TwoWire() Can't init
............
[11:40:06][I][application:113]: Running through first loop()
[11:40:06][E][esp32-hal-i2c.c:1146] i2cCheckLineState(): Bus Invalid State, TwoWire() Can't init
[11:40:06][W][i2c:093]: Unknown transmit error 5 for address 0x5A
[11:40:06][I][application:140]: First loop finished successfully!
[11:40:06][E][esp32-hal-i2c.c:1146] i2cCheckLineState(): Bus Invalid State, TwoWire() Can't init
[11:40:06][W][i2c:093]: Unknown transmit error 5 for address 0x5A
[11:40:06][E][esp32-hal-i2c.c:1146] i2cCheckLineState(): Bus Invalid State, TwoWire() Can't init
[11:40:06][W][i2c:093]: Unknown transmit error 5 for address 0x5A
```

As I mentioned before with the dev version of arduino esp32 The mpr121 works fine on esp32 as well.

```yaml
esphome:
  name: esp32mpr121
  platform: ESP32
  board: nodemcu-32s
  arduino_version: dev
```

**Review by OttoWinter — APPROVED** [review]

**Comment by germil** [comment]
Hello, please help me with advice on how to change a treshold - sensitivity for sensors.
Thank you, 
and all congrats for your great job here.

**Comment by mvturnho** [comment]
> Hello, please help me with advice on how to change a treshold - sensitivity for sensors.
> Thank you,
> and all congrats for your great job here.

Hi, this should be changed in the driver code. For now the values for touch- and release-tresholds are fixed:
touch 12
release 6

I might add this at some moment.

**Comment by mvturnho** [comment]
> Hello, please help me with advice on how to change a treshold - sensitivity for sensors.
> Thank you,
> and all congrats for your great job here.

I have added support for debounce and threshold to the sensor. You can test it with the folowing yaml:
NOTE: you need to use the esphome_core_version in the yaml below or it will not work.

```yaml
esphome:
  name: mpr121
  platform: ESP8266
  board: nodemcuv2
  esphome_core_version:
    repository: https://github.com/mvturnho/esphome-core.git
    branch: mpr121

# Enable logging
logger:
  level: DEBUG

i2c:
  sda: D2
  scl: D1
  scan: True
  frequency: 100kHz
  
mpr121:
#  id: mpr121_first
  address: 0x5A
  touch_debounce: 7
  release_debounce: 3
  
binary_sensor:
  - platform: mpr121
    channel: 0
    id: touchkey0
    touch_threshold: 0x30
    release_threshold: 0x30
  - platform: mpr121
    channel: 1
    id: touchkey1
    touch_threshold: 12
    release_threshold: 6
  - platform: mpr121
    channel: 2
    id: touchkey2
  - platform: mpr121
    channel: 3
    id: touchkey3
    touch_threshold: 5
    release_threshold: 5
  - platform: mpr121
    channel: 4
    id: touchkey4
  - platform: mpr121
    channel: 11
    id: touchkey11
    touch_threshold: 0x30
    release_threshold: 0x30
```

**Comment by germil** [comment]
I cant get it to work. When i load just part of configuration i get that log:

`
INFO Reading configuration...
INFO Configuration is valid!
esphome:
  name: mpr121
  platform: ESP8266
  board: nodemcuv2
  esphome_core_version:
    repository: https://github.com/mvturnho/esphome-core.git
    branch: mpr121
  build_path: mpr121
  arduino_version: espressif8266@1.8.0
  use_custom_code: false
  board_flash_mode: dout
logger:
  level: DEBUG
  baud_rate: 115200
  tx_buffer_size: 512
  hardware_uart: UART0
i2c:
  sda: 4
  scl: 5
  scan: true
  frequency: 100000.0
`

log:
`
INFO Reading configuration...
INFO Starting log output from tipke.local using esphome API
INFO Connecting to tipke.local:6053 (192.168.1.205)
INFO Successfully connected to tipke.local
[21:35:28][I][application:097]: esphome-core version 1.13.0-dev compiled on Apr  1 2019, 21:21:03
[21:35:28][C][wifi:371]: WiFi:
[21:35:28][C][wifi:253]:   SSID: [redacted]
[21:35:28][C][wifi:254]:   IP Address: 192.168.1.205
[21:35:28][C][wifi:256]:   BSSID: [redacted]
[21:35:28][C][wifi:257]:   Hostname: 'tipke'
[21:35:28][C][wifi:261]:   Signal strength: -38 dB ▂▄▆█
[21:35:28][C][wifi:262]:   Channel: 6
[21:35:28][C][wifi:263]:   Subnet: 255.255.255.0
[21:35:28][C][wifi:264]:   Gateway: 192.168.1.1
[21:35:29][C][wifi:265]:   DNS1: 192.168.1.1
[21:35:29][C][wifi:266]:   DNS2: 0.0.0.0
[21:35:29][C][i2c:035]: I2C Bus:
[21:35:29][C][i2c:036]:   SDA Pin: GPIO4
[21:35:29][C][i2c:037]:   SCL Pin: GPIO5
[21:35:29][C][i2c:038]:   Frequency: 100000 Hz
[21:35:29][I][i2c:040]: Scanning i2c bus for active devices...
[21:35:29][I][i2c:047]: Found i2c device at address 0x5A
[21:35:29][C][logger:142]: Logger:
[21:35:29][C][logger:143]:   Level: DEBUG
[21:35:29][C][logger:144]:   Log Baud Rate: 115200
[21:35:29][C][logger:145]:   Hardware UART: UART0
[21:35:29][C][binary_sensor.mpr121:080]: MPR121:
[21:35:29][C][binary_sensor.mpr121:081]:   Address: 0x5A
[21:35:29][C][api:101]: API Server:
[21:35:29][C][api:102]:   Address: tipke.local:6053
[21:35:29][C][ota:127]: Over-The-Air Updates:
[21:35:29][C][ota:128]:   Address: tipke.local:8266`

But when i add touch and relese debounce:

`esphome:
  name: tipke
  platform: ESP8266
  board: d1_mini
  esphome_core_version:
    repository: https://github.com/mvturnho/esphome-core.git
    branch: mpr121

wifi:
  ssid: "CISCOMG"
  password: "xxxxxxxxxxxx"

api:

ota:

logger:
  level: DEBUG

i2c:
  sda: D2
  scl: D1
  scan: True
  frequency: 100kHz
  
mpr121:
  id: mpr121_first
  address: 0x5A
  touch_debounce: 7
  release_debounce: 3
  
binary_sensor:
  - platform: mpr121
    channel: 0
    id: touchkey0
    touch_threshold: 0x30
    release_threshold: 0x30
  - platform: mpr121
    channel: 1
    id: touchkey1
    touch_threshold: 12
    release_threshold: 6
  - platform: mpr121
    channel: 2
    id: touchkey2
  - platform: mpr121
    channel: 3
    id: touchkey3
    touch_threshold: 5
    release_threshold: 5
  - platform: mpr121
    channel: 4
    id: touchkey4
  - platform: mpr121
    channel: 11
    id: touchkey11
    touch_threshold: 0x30
    release_threshold: 0x30`

i get that error log:

`INFO Reading configuration...
Failed config

mpr121: 
  - [source /config/esphome/tipke.yaml:28]
    id: mpr121_first
    address: 90
    
    [touch_debounce] is an invalid option for [mpr121]. Please check the indentation.
    touch_debounce: 7
    
    [release_debounce] is an invalid option for [mpr121]. Please check the indentation.
    release_debounce: 3
binary_sensor.mpr121: [source /config/esphome/tipke.yaml:34]
  platform: mpr121
  channel: 0
  id: touchkey0
  
  [touch_threshold] is an invalid option for [binary_sensor.mpr121]. Please check the indentation.
  touch_threshold: 48
  
  [release_threshold] is an invalid option for [binary_sensor.mpr121]. Please check the indentation.
  release_threshold: 48
binary_sensor.mpr121: [source /config/esphome/tipke.yaml:39]
  platform: mpr121
  channel: 1
  id: touchkey1
  
  [touch_threshold] is an invalid option for [binary_sensor.mpr121]. Please check the indentation.
  touch_threshold: 12
  
  [release_threshold] is an invalid option for [binary_sensor.mpr121]. Please check the indentation.
  release_threshold: 6
binary_sensor.mpr121: [source /config/esphome/tipke.yaml:47]
  platform: mpr121
  channel: 3
  id: touchkey3
  
  [touch_threshold] is an invalid option for [binary_sensor.mpr121]. Please check the indentation.
  touch_threshold: 5
  
  [release_threshold] is an invalid option for [binary_sensor.mpr121]. Please check the indentation.
  release_threshold: 5
binary_sensor.mpr121: [source /config/esphome/tipke.yaml:55]
  platform: mpr121
  channel: 11
  id: touchkey11
  
  [touch_threshold] is an invalid option for [binary_sensor.mpr121]. Please check the indentation.
  touch_threshold: 48
  
  [release_threshold] is an invalid option for [binary_sensor.mpr121]. Please check the indentation.
  release_threshold: 48`

**Comment by mvturnho** [comment]
Correct. You also need to update your esphome to my mpr121 branch.

I think it was
pip install --upgrade git+https://github.com/mvturnho/esphome@mpr121

then it recognises the new config in the yaml file.

Also could you comment in that repo until we have a working version. I don't know if this discussion should be held here. This PR was already closed in the previous version.

**Pull request merged**